images/sds-replicated-volume-controller/src/pkg/controller/replicated_storage_class.go

```diff
@@ -515,49 +583,77 @@ func GetStorageClass(ctx context.Context, cl client.Client, namespace, name stri
 }
 
 func DeleteStorageClass(ctx context.Context, cl client.Client, sc *storagev1.StorageClass) error {
-	return cl.Delete(ctx, sc)
-}
-
-func RemoveString(slice []string, s string) (result []string) {
-	for _, value := range slice {
-		if value != s {
-			result = append(result, value)
+	finalizers := sc.ObjectMeta.Finalizers
+	switch len(finalizers) {
+	case 0:
+		return cl.Delete(ctx, sc)
+	case 1:
+		if finalizers[0] != StorageClassFinalizerName {
+			return fmt.Errorf("deletion of StorageClass with finalizer %s is not allowed", finalizers[0])
 		}
+		sc.ObjectMeta.Finalizers = nil
+		if err := cl.Update(ctx, sc); err != nil {
+			return fmt.Errorf("error updating StorageClass to remove finalizer %s: %w",
+				StorageClassFinalizerName, err)
+		}
+		return cl.Delete(ctx, sc)
 	}
-	return
+	// The finalizers list contains more than one element — return an error
+	return fmt.Errorf("deletion of StorageClass with multiple(%v) finalizers is not allowed", finalizers)
 }
 
-func ReconcileStorageClassLabelsAndAnnotationsIfNeeded(ctx context.Context, cl client.Client, oldSC, newSC *storagev1.StorageClass) error {
-	if !reflect.DeepEqual(oldSC.Labels, newSC.Labels) || !reflect.DeepEqual(oldSC.Annotations, newSC.Annotations) {
-		oldSC.Labels = newSC.Labels
-		oldSC.Annotations = newSC.Annotations
-		return cl.Update(ctx, oldSC)
+func updateStorageClassMetaDataIfNeeded(
+	ctx context.Context,
+	cl client.Client,
+	newSC, oldSC *storagev1.StorageClass,
+) error {
+	needsUpdate := !reflect.DeepEqual(oldSC.Labels, newSC.Labels) ||
+		!reflect.DeepEqual(oldSC.Annotations, newSC.Annotations) ||
+		!reflect.DeepEqual(oldSC.Finalizers, newSC.Finalizers)
```

Comment:
There's a case, when the only difference between finalizers is their order. It's possible, because we don't keep the order when copying them in `DoUpdateStorageClass`.

I think the right way to do it is to order-agnostic comparison here.

Reply:
good point